File: demo/src/com/novoda/merlin/demo/presentation/base/MerlinActivity.java
package com.novoda.merlin.demo.presentation.base;

import android.app.Activity;
import android.os.Bundle;

import com.novoda.merlin.Merlin;
import com.novoda.merlin.registerable.bind.Bindable;
import com.novoda.merlin.registerable.disconnection.Disconnectable;
import com.novoda.merlin.registerable.connection.Connectable;

public abstract class MerlinActivity extends Activity {

    private Merlin merlin;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        merlin = createMerlin();
    }

    protected abstract Merlin createMerlin();

    @Override
    protected void onResume() {
        super.onResume();
        merlin.bind();
    }

    @Override
    protected void onPause() {
        super.onPause();
        merlin.unbind();
    }

    protected void registerConnectable(Connectable connectable) {
        merlin.registerConnectable(connectable);
    }

    protected void registerDisconnectable(Disconnectable disconnectable) {
        merlin.registerDisconnectable(disconnectable);
    }

    protected void registerBindable(Bindable bindable) {
        merlin.registerBindable(bindable);
    }

}

File: demo/src/main/java/com/novoda/merlin/demo/presentation/base/MerlinActivity.kt
package com.novoda.merlin.demo.presentation.base

import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.novoda.merlin.Merlin
import com.novoda.merlin.logger.Logger
import com.novoda.merlin.registerable.bind.Bindable
import com.novoda.merlin.registerable.connection.Connectable
import com.novoda.merlin.registerable.disconnection.Disconnectable

abstract class MerlinActivity() : AppCompatActivity() {

    private val logHandle = DemoLogHandle()
    protected abstract val merlin: Merlin

    protected fun registerConnectable(connectable: Connectable) {
        merlin.registerConnectable(connectable)
    }

    protected fun registerDisconnectable(disconnectable: Disconnectable) {
        merlin.registerDisconnectable(disconnectable)
    }

    protected fun registerBindable(bindable: Bindable) {
        merlin.registerBindable(bindable)
    }

    override fun onStart() {
        super.onStart()
        Logger.attach(logHandle)
        merlin.bind()
    }

    override fun onStop() {
        super.onStop()
        merlin.unbind()
        Logger.detach(logHandle)
    }

    class DemoLogHandle : Logger.LogHandle {

        companion object {
            const val TAG = "DemoLogHandle"
        }

        override fun v(vararg message: Any?) {
            Log.v(TAG, message[0].toString())
        }

        override fun i(vararg message: Any?) {
            Log.i(TAG, message[0].toString())
        }

        override fun d(vararg msg: Any?) {
            Log.d(TAG, msg[0].toString())
        }

        override fun d(throwable: Throwable?, vararg message: Any?) {
            Log.d(TAG, message[0].toString(), throwable)
        }

        override fun w(vararg message: Any?) {
            Log.w(TAG, message[0].toString())
        }

        override fun w(throwable: Throwable?, vararg message: Any?) {
            Log.w(TAG, message[0].toString(), throwable)
        }

        override fun e(vararg message: Any?) {
            Log.e(TAG, message[0].toString())
        }

        override fun e(throwable: Throwable?, vararg message: Any?) {
            Log.e(TAG, message[0].toString(), throwable)
        }

    }

}
